rfcs/0024-nondeterministic-functions.md

```diff
@@ -0,0 +1,60 @@
+---
+feature: non_deterministic_functions
+authors:
+  - "st1page"
+start_date: "2022/11/21"
+---
+
+# nondeterministic-functions
+
+## Summary
+
+We introduce the `MaterializeProject` stream executor to handle non-deterministic function on updatable stream.
+
+## Motivation
+
+In the discussion about the udf, @skyzh raise the issue about the non-deterministic function on updatable stream <https://github.com/risingwavelabs/rfcs/pull/12#issuecomment-1305105614>. This issue also appears on the built-in functions such as `random(),` `proc_time()` or `now()`.
+
+## Design
+
+We will introduce a `MaterializeProject`, it will materialize some **partial** columns of result with the stream key of the input as the primiary key. When an `Insert` operation comes, it will compute the result and materialize some columns. when `Update` or `Delete` comes, it will lookup its state to replace the old value of the operation.
+For example, if we have a materialized view with SQL `select random() as rd, v*2 as vv, pk from t` and the pk is the primary key of the table t. The stream processing plan will be like:
+
+```plain
+StreamMaterialize(columns:[rd, vv, pk])
+  StreamMaterializeProject(
+      exprs: [Random(), InputRef(0) * 2, InputRef(1)],
+      materialized_cols: [0],
+      state_table: [pk, rd]
+    )
+    StreamScan(fields:[v, pk])
+```
+
+The `StreamMaterializeProject` will materialize the first columns of its input which is the `random()` function's result. When we have a serial of changes `Insert(v = 10, pk = 1), Update((v = 10, pk = 1) -> (v = 20, pk = 1)), Delete(v = 20, pk = 1)` the `StreamMaterializeProject` will handle it as following:
```

Comment:
The `StreamMaterializeProject` will materialize all output of projection with non-deterministic result.